### ISSUE TITLE
warning when using one string at the end a module to return true

### ISSUE THREAD

**Issue by atoomic** [Perl/perl5#17982]
When using `warnings` and a different value than `1` at the end of a perl module, this raises a `Useless use of a constant in void context` warnings.

It appears that we should not raise such warnings in this context.
I would even suggest that we should keep the warning on and rephrase it to something better.
IMO, using something like `"1"x100_000_000` (or maybe shorter) seems something we should not try to encourage at the end of a module.

Note: We started a discussion as part of PR #17981 but we should isolate it to its own thread
copy: @Leont @hvds 

As
```
> echo 'package Foo; "abcd"' > Foo.pm && perl -cw Foo.pm
Useless use of a constant ("abcd") in void context at Foo.pm line 1.
Foo.pm syntax OK
```

In truth using any integer should not raise a warning.
```
> echo 'package Foo; 12345' > Foo.pm && perl -cw Foo.pm
Useless use of a constant (12345) in void context at Foo.pm line 1.
Foo.pm syntax OK
```

```
> echo 'package Foo; 2' > Foo.pm && perl -cw Foo.pm
Useless use of a constant (2) in void context at Foo.pm line 1.
Foo.pm syntax OK
```

no warnings when using `1` only
```
> echo 'package Foo; 1' > Foo.pm && perl -cw Foo.pm
Foo.pm syntax OK
```

Maybe in this case `0` should also return a warning.
```
> echo 'package Foo; 0' > Foo.pm && perl -cw Foo.pm
Foo.pm syntax OK
```

**Comment by hvds** [comment]
My two cents: it should not warn at all. Docs for require state only that it will raise an error if not a true value, so it is also not required to be an integer.

**Comment by Grinnz** [comment]
Two parts:

1. I would guess that the warning occurs because the result of the module, returned by require, is called in void context (as require almost always is). "1" is specifically exempt from this warning in all cases, that's why it doesn't warn here. It would be nice to have the warning not fire in this case, regardless of value.

2. I see no reason to encourage or discourage what the final true value of a module should be with a new warning, especially if we plan to make it unnecessary eventually (#17921).

**Comment by atoomic** [comment]
Indeed after #17921 it makes no sense to encourage or not on the final value of a module.

We should avoid such a warning when loading a module

**Comment by Grinnz** [comment]
I also would point out that the only reason your examples encounter this warning is because you are "running" the module directly, rather than via a require. `perl -w -mFoo -e1` does not warn.

**Comment by Leont** [comment]
> I also would point out that the only reason your examples encounter this warning is because you are "running" the module directly, rather than via a require. perl -w -mFoo -e1 does not warn.

Yeah, that is an important thing to note.

**Comment by tonycoz** [comment]
Does Module::CoreList::Utils warn during the test run with default warnings?  (which the original PR seems to imply but does not demonstrate.)

**Comment by atoomic** [comment]
No it does warn during the test run, as described in the Pull Request the warnings occurs when trying to run a `perl -cw` on the module itself

```
> perl -cw lib/Module/CoreList/Utils.pm
Useless use of a constant ("foo") in void context at lib/Module/CoreList/Utils.pm line 1665.
lib/Module/CoreList/Utils.pm syntax OK
```

I understand that running it like `perl -w -Ilib -mModule::CoreList::Utils -e1` is not raising warnings and the problem comes from the require loading the module.

I agree no warnings should be raised when loading a .pm file under that context.
I think running a `perl -cw` on a module file, is a common usage, that we should accommodate.

**Comment by Grinnz** [comment]
> I think running a `perl -cw` on a module file, is a common usage, that we should accommodate.

I strongly disagree on both counts.

**Comment by Grinnz** [comment]
> No it does warn during the test run, as described in the Pull Request the warnings occurs when trying to run a `perl -cw` on the module itself

This is not a test, this is just something you decided to do incorrectly.

**Comment by tonycoz** [comment]
> I think running a `perl -cw` on a module file, is a common usage, that we should accommodate.

I'm confused at what the issue is.

An author can create a .pm that loads with both `perl -cw Foo.pm` and `perl -MFoo -e0` by using `1;` as the final line of their module.

For your test case perl is correctly reporting that a more complex constant like `"My Novel...";` is being evaluated in void context when run with `perl -cw`, since the value isn't actually being used.  For `perl -MFoo` aka `require Foo` the value is used.

What fix are you looking for?

**Comment by atoomic** [comment]
I would have expect that `perl -cw Module.pm` would not raise a warning when using a different value than `1` at the end of a file.

But it appears that:
- this is an incorrect usage of `perl -cw`
- returning `true value` requirement might be removed

As this is clearly not a new behavior. This is a misunderstanding from my side.
I think we should close this case without any actions.

thanks all for your valuable input
